README.txt
RippleSite version 0.2 README 
----------------------------- 

by Ryan Fugger
ryan@ripplepay.com 
http://ripple.sourceforge.net/

License 
------- 

The RippleSite software package, including source code, HTML templates, 
documentation, artwork, and all other included files, is covered by the 
GNU General Public License, a copy of which can be found in LICENSE.txt 
in the same directory as this README.  

-- 
RippleSite Standalone Ripple Web Service Software
Copyright (C) 2006-07

This program is free software; you can redistribute it and/or modify it 
under the terms of the GNU General Public License as published by the 
Free Software Foundation; either version 2 of the License, or (at your 
option) any later version.

This program is distributed in the hope that it will be useful, but 
WITHOUT ANY WARRANTY; without even the implied warranty of 
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU General 
Public License for more details.

You should have received a copy of the GNU General Public License along 
with this program; if not, write to the Free Software Foundation, Inc., 
59 Temple Place, Suite 330, Boston, MA 02111-1307 USA
--

Description 
----------- 

RippleSite is an open source version of the server software used to run 
RipplePay.com, a free web service for hosting Ripple payment networks on 
a single server. (For more information about the Ripple payment concept, 
see http://ripple.sourceforge.net/.) To get RippleSite running on your 
web server, you need a fair amount of Python, Apache, and PostgreSQL 
know-how. 

The Django project website has a terrific tutorial that anyone 
who intends to work on RippleSite as a developer should go through. 

Installation 
------------ 

RippleSite 0.3 works on the head version of the Django Python web application framework.
You need to check this out of subversion and install as described at 

  http://www.djangoproject.com/download/  

RippleSite requires the following software: 

* root acces on some unix computer. I use ubuntu, on a $20/month rented virtual host from linode.com
  it may be possible to get ripple to work on a shared hosting provider such as dreamhost
  where you don't have root access, but likely to be extremely difficult and not worth the effort.
* Python (I'm using 2.4) 
* Python modules
    -- django, head (I installed via svn co http://code.djangoproject.com/svn)
    -- psycopg (I installed via apt-get python-psycopg)

* Apache + mod_python (I'm using Apache 2.0), 
  It is recommended to run the django development server to try it out first, as this is easier than apache setup
  on ubuntu, apt-get install apache2 libapache2-mod-python

* PostgreSQL (tested on 8.0, 8.1, 8.2)
    I installed via apt-get install postgresql
* an smtp mail server for sending signup confirmations.
  I am using gmail can be used as an smtp mail server, there is an example of this in settings.py


0. Basic preparation
   Check out the source code containing this readme file.
   Install python, django head from svn, postgresql, smtp server (which can just be a gmail account)

  On a debian/ubuntu box,

  sudo apt-get install python postgres python-psycopg 

1. Edit settings.py for passwords, domain names, etc unique to your instance of ripple

2. Create the ripplesite postgres database, and a postgres user that matches your system user

  run ./postgres-setup.py 

  This assumes postgres superuser is named postgres, as is standard, and you have sudo installed,
  and have the right to run commands as a superuser via sudo.
       
  run 'python manage.py syncdb'   # initialize the ripplesite database. 

    create a django app superuser/password as the questions direct you to.

4. Configure Apache.

    (You can skip this if you only want to run the django development server.
    The django development server should be fine unless you have a lot of users.)

    Put something like the following in httpd.conf: 

    -- 
    <Location "/"> 
      SetHandler python-program 
      PythonHandler django.core.handlers.modpython 
      SetEnv DJANGO_SETTINGS_MODULE ripplesite.settings
      PythonPath "['/path/to/parent'] + sys.path" 
      PythonDebug Off 
    </Location> 

    <Location "/media/">
      SetHandler None 
    </Location> 
    -- 

    * /path/to/parent is the path to the parent directory of the
    ripplesite project directory

    You must also create a symbolic link to the admin media files from 
    within your Apache document root. There is more information on how to 
    configure Apache for Django + mod_python at:

    http://www.djangoproject.com/documentation/modpython/

4.5 Instructions for running the django development server

  run ./run-server.sh

5. Browse to

     http://mywebhost.com/admin/ (apache)
     or http://mywebhost.com:8080/admin (django dev server)


   log into the Django admin app using the password you created during the syncdb step earlier.
   Click on Currency units and create 
   the currency units you'd like your users to be able to choose on your 
   site. If you'd like to use US dollars, for example, create a currency 
   unit with short name 'USD', long name 'US Dollar', symbol '$', and value 
   1.0. You can always create more later, but you'll need at least one to 
   start with. 

   If you have multiple units that you want Ripple to be able to 
   automatically convert between, set their relative values in the value 
   fields. For example, if you had US dollars set to a value of 1.0, 
   Canadian dollars would be set to around 0.90 because one dollar Canadian 
   is currently worth about 90 cents US. (I've included the script 
   getrates.py in the scripts directory which will fetch current exchange 
   rates from Yahoo and output SQL statements for updating the database -- 
   that should be a decent starting point for piecing together an 
   auto-update system.) 

6. Browse to the root on your site and create a new user. Send out some 
invitations to people you know... 

  htpp://mywebhost.com (apache)
  http://mywebhost.com:8080 (django dev server)

7. You can customize your home page by modifying the file 
ripple/templates/home.html. 

8. You may want to set up a forum where your users can find items they 
can buy with Ripple. Suggestions are a web forum, a Yahoo group, a 
Google group, or Craigslist. In the case of the latter two public 
forums, if you direct your users to include a special prefix such as 
*VanRipple* (if your service was called VanRipple) with every post, they 
will be able to search specifically for Ripple items. 

9. If you expect a lot of users, remember you can get a lot better 
performance by creating the indexes generated by python manage.py 
sqlindexes and tuning your PostgreSQL installation. See 
http://www.revsys.com/writings/postgresql-performance.html. 

  Note: This documentation may be out of date.
        I think django head does indexes automatically
        (Double check this before deleting point 9 though.)

Contact Info 
------------ 

If you need help getting your Ripple site set up, have a feature request 
to make, or better yet, have some improvements to the code that you wish 
to contribute to the project (I should be so lucky), please post to the 
Ripple project forums at https://sourceforge.net/forum/?group_id=120019. 
I am also currently (Feb. 2007) working on a P2P version of Ripple to 
eventually be integrated with RippleSite, if anyone would like to help. 

Ideas for Improvements 
---------------------- 

If you feel like improving the code, here are some ideas for things to 
work on: 

* improve the site design and layout (I'm not a web designer)
* write an installation script 
* integrate admin-driven news items into the front page 
* paginate payment history in both account and payment views 
* allow user to request payment from other users 
* allow multiple accounts between two users (may simply be a case of 
removing the check for this when creating new account and testing 
that it works with payments) 
* allow users to view current conversion rates 
* allow users to print up certificates and/or generate codes that may 
be given to other users and redeemed on the system, for face-to-face 
transactions 
* translations into other languages 
* make email notifications optional and/or encrypt them with user's 
PGP key 
* security question for resetting password 
* longer confirmation/forgot password codes 
* credit/reputation checks - request permission from user to find out 
maximum amount they could pay you 
* contact form on contact page 
* improve test-cookie login handling (these test sessions build up and 
have to be cleared periodically) 
* change account units 
* transfer between accounts 
* invite-only registrations 
* resend invitation or receipt 
* split off user/login info from Node model and use builtin Django User 
instead 



CHANGES.txt
Changes in RippleSite 0.2
-------------------------

* ability to charge interest on account balances
* proper escaping to prevent XSS attacks


ripple/models.py
# RippleSite Standalone Ripple Web Service Software
# Copyright (C) 2006-07
# 
# This program is free software; you can redistribute it and/or modify it 
# under the terms of the GNU General Public License as published by the 
# Free Software Foundation; either version 2 of the License, or (at your 
# option) any later version.
#
# This program is distributed in the hope that it will be useful, but 
# WITHOUT ANY WARRANTY; without even the implied warranty of 
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU General 
# Public License for more details.
#
# You should have received a copy of the GNU General Public License along 
# with this program; if not, write to the Free Software Foundation, Inc., 
# 59 Temple Place, Suite 330, Boston, MA 02111-1307 USA
#
# Ryan Fugger
# ryan@ripplepay.com
#-------------------------------------------------------------------------

from django.db import models

import sha
from datetime import datetime

#todo: validators?
#todo: add admin tags, default values, etc. - nice model things
#todo: split Node up into User and Node with 1-1 relationship to speed up node fetching??

DIGITS = 24 # max. number of digits in amounts
DECIMALS = 12 # digits reserved for decimals

class CurrencyUnit(models.Model):
    short_name = models.CharField(max_length=5, unique=True)
    long_name = models.CharField(max_length=50)
    symbol = models.CharField(max_length=3)
    value = models.FloatField(default=0.0) # in base units - for conversion
    last_update = models.DateTimeField(auto_now=True) # last value update
    
    class Admin:
        list_display = ('long_name', 'short_name', 'symbol', 'value', 'last_update')
    
    class Meta:
        ordering = ('short_name',)
    
    def __str__(self):
        return self.short_name


class Node(models.Model):
    username = models.CharField(max_length=20, unique=True)
    name = models.CharField(max_length=50)
    pwd = models.CharField(max_length=40)
    salt = models.CharField(max_length=10, default='') # for salting the password before hashing
    display_units = models.ForeignKey(CurrencyUnit, null=True, blank=True, default=None)
    do_conversions = models.BooleanField(default=True) # allow conversion on through payments
    #notify_emails = models.BooleanField()
    #bulletins = models.BooleanField()
    location = models.FloatField() # routing location

    class Admin:
        list_display = ('username', 'name', 'display_units', 'do_conversions')
        list_filter = ('display_units', 'do_conversions')
        search_fields = ('username', 'name')
    
    def __str__(self):
        return self.name or self.username
    
    def checkPwd(self, pwd):
        if self.pwd == sha.new(pwd + self.salt).hexdigest():
            return True
        else: return False
    
    def setPwd(self, pwd):
        self.pwd = sha.new(pwd + self.salt).hexdigest()
    
    def getPrimaryEmail(self):
        return self.emailaddr_set.get(primary__exact=True).email


class EmailAddr(models.Model):
    node = models.ForeignKey(Node, db_index=True)
    email = models.EmailField(unique=True)
    code = models.CharField(max_length=10, unique=True)
    confirmed = models.BooleanField(default=False)
    primary = models.BooleanField()
    
    class Admin:
        list_display = ('email', 'node', 'confirmed', 'primary')
        list_filter = ('confirmed', 'primary')
        search_fields = ('email',)

    def __str__(self):
        return self.email


class Reminder(models.Model):
    node = models.ForeignKey(Node)
    code = models.CharField(max_length=10, unique=True)
    date = models.DateTimeField(auto_now_add=True)
    
    def __str__(self):
        return "%s %s" % (self.node, self.date)

    class Admin:
        list_display = ('node', 'code', 'date')


# accounts ----------------------------------
class SharedAccountData(models.Model):
    currency = models.ForeignKey(CurrencyUnit)
    balance = models.FloatField()
    active = models.BooleanField(default=True)
    created = models.DateTimeField(auto_now_add=True)
    
    # interest
    interest_rate = models.FloatField(default=0.0) # stored as decimal (as opposed to percentage for display)
    proposed_rate = models.FloatField(null=True, blank=True, default=None)
    node_to_confirm = models.ForeignKey(Node, related_name='shared_accounts_to_confirm', null=True, blank=True, default=None)
    last_update = models.DateTimeField(auto_now_add=True) # to quickly compute present interest tally during payments
    # (interest is added to balance each update, ie, each transaction)

    class Admin:
        list_display = ('currency', 'balance', 'active', 'interest_rate', 'last_update')
        list_filter = ('currency', 'active')
    
    def __str__(self):
        str = '%s account %d' % (self.currency, self.id)
        if not self.active:
            str += " (inactive)"
        return str
    
    def displayRate(self):
        return self.interest_rate * 100.0
    def displayProposedRate(self):
        if self.proposed_rate <> None:
            return self.proposed_rate * 100.0
        return None
    
    # avoid rounding on save when django converts to string for SQL command
    def save(self):
        if type(self.balance) == float:
            self.balance = '%.12f' % self.balance
        super(SharedAccountData, self).save()
        self.balance = float(self.balance)

class Account(models.Model):
    owner = models.ForeignKey(Node, db_index=True)
    partner = models.ForeignKey(Node, related_name='partner_account_set', db_index=True)
    shared_data = models.ForeignKey(SharedAccountData)
    balance_multiplier = models.IntegerField(choices=[(1, '+1'), ( -1, '-1')])
    #partner_acct = models.ForeignKey("self", null=True, blank=True)

    # what owner calls the account (default would be name/email of partner)
    name = models.CharField(max_length=50)

    # current max negative (debt) balance of this account
    iou_limit = models.FloatField()

    # meta-limits - acceptable range for upper and lower account limits according to owner
    # used to allow each partner to regulate their own credit flexibly
    # iou_limit = min(my_limit, partner_acct.partner_limit)
    partner_limit = models.FloatField() # max amt to accept from partner
    my_limit = models.FloatField(null=True, blank=True) # max amt of debt to owe to partner - null -> no limit

    class Admin:
        list_display = ('id', 'name', 'owner', 'shared_data', 'iou_limit', 'partner_limit', 'my_limit')
        search_fields = ('name', 'owner__name', 'owner__username', 'partner__name', 'partner__username')

    def __str__(self):
        return '%s\'s view of %s, "%s"' % (self.owner, self.shared_data, self.name)

    def get_partner_acct(self):
        return Account.objects.get(shared_data__pk=self.shared_data_id, owner__pk=self.partner_id)
    
    def getAvailableCredit(self):
        return self.iou_limit + self.getBalance() #- frozen

    def getBalance(self):
        return self.getBalanceNoInterest() + self.getPresentInterest()

    def getBalanceNoInterest(self):
        return self.shared_data.balance * self.balance_multiplier

    def getPresentInterest(self):
        from ripplesite.ripple.views.pathgraph import computeInterest
        return computeInterest(self.getBalanceNoInterest(), self.shared_data.interest_rate, self.shared_data.last_update, datetime.now())


class Offer(models.Model):
    initiator = models.ForeignKey(Node, related_name='sent_offer_set')
    recipient_email = models.CharField(max_length=50)
    amount = models.FloatField()
    currency = models.ForeignKey(CurrencyUnit)
    initial_balance = models.FloatField(default=0.0)
    interest_rate = models.FloatField() # stored as percentage, 2 fewer decimals

    class Admin:
        list_display = ('initiator', 'recipient_email', 'amount', 'currency', 'initial_balance')
        list_filter = ('currency',)
        search_fields = ('initiator__name', 'initiator__username', 'recipient_email')

    def __str__(self):
        return "%s offer %s %.2f to %s" % (self.initiator, self.currency, self.amount, self.recipient_email)


# payments ---------------------------
PAYMENT_STATUS_CHOICES = (
    ('RQ', 'Requested'),
    ('PE', 'Pending'),
    ('OK', 'Completed'),
    ('CA', 'Cancelled'),
    ('RF', 'Refused'),
    ('FA', 'Failed'),
)

class Payment(models.Model):
    payer = models.ForeignKey(Node, related_name='outgoing_payment_set', db_index=True)
    payer_email = models.EmailField()
    recipient = models.ForeignKey(Node, related_name='incoming_payment_set', db_index=True)
    recipient_email = models.EmailField()
    date = models.DateTimeField(auto_now_add=True)
    currency = models.ForeignKey(CurrencyUnit)
    amount = models.FloatField()
    status = models.CharField(max_length=2, choices=PAYMENT_STATUS_CHOICES)
    description = models.TextField(default='', blank=True)

    def __str__(self):
        return "%d: %s pay %s %.2f to %s" % (self.id, self.payer, self.currency, self.amount, self.recipient)

    class Admin:
        list_display = ('id', 'payer', 'recipient', 'date', 'currency', 'amount', 'status')
        list_filter = ('currency', 'status')
        search_fields = ('payer__name', 'payer__username', 'payer_email', 'recipient__name', 'recipient__username', 'recipient_email', 'description')
    
    # avoid rounding on save when django converts to string for SQL command
    def save(self):
        if type(self.amount) == float:
            self.amount = '%.12f' % self.amount
        super(Payment, self).save()
        self.amount = float(self.amount)
        

# payment may have multiple paths
class PaymentPath(models.Model):
    payment = models.ForeignKey(Payment)
    amount = models.FloatField()

    def __str__(self):
        return "%d: %.2f on payment %s" % (self.id, self.amount, self.payment)
    
    # avoid rounding on save when django converts to string for SQL command
    def save(self):
        if type(self.amount) == float:
            self.amount = '%.12f' % self.amount
        super(PaymentPath, self).save()
        self.amount = float(self.amount)


# each link is an account from the perspective of the partner closest to payment recipient
# (the IOU recipient for this link)
class PaymentLink(models.Model):
    path = models.ForeignKey(PaymentPath)
    payer_account = models.ForeignKey(Account)
    position = models.PositiveIntegerField() # position in path sequence, starting at 1

    # needed because this account may be in a different currency than payment
    amount = models.FloatField()
    interest = models.FloatField() # add this to amount to get amount actually subtracted from acct balance during transaction
    interest_rate = models.FloatField() # for auditability

    def __str__(self):
        return "%d. %s on path %s" % (self.position, self.payer_account, self.path)
    
    # avoid rounding on save when django converts to string for SQL command
    def save(self):
        if type(self.amount) == float:
            self.amount = '%.12f' % self.amount
        if type(self.interest) == float:
            self.interest = '%.12f' % self.interest
        super(PaymentLink, self).save()
        self.amount = float(self.amount)
        self.interest = float(self.interest)

